Move pulse_wigner to LaserPulse class

danielhollas opened the pull request:
This is imo the last method that should obviously be moved  to LaserPulse class.

danielhollas commented on src/promdens/promdens.py:
```diff
@@ -163,6 +163,48 @@ def calc_field_envelope(self, t: np.ndarray) -> np.ndarray:
+        """
+        # setting an adaptive integration step according to the frequency of the integrand oscillations (de - omega)
+        loc_omega = self.omega + 2*self.lchirp*tprime
+        # If de == loc_omega, we're in resonance and there are no oscillations,
+        # integrate only the envelope intensity so we can set a larger integration step
```
I've expanded this comment a bit, let me know if it makes sense.

danielhollas commented on src/promdens/promdens.py:
```diff
@@ -163,6 +163,48 @@ def calc_field_envelope(self, t: np.ndarray) -> np.ndarray:
+        if de == loc_omega:
+            T = 2*np.pi/loc_omega
+        else:
+            T = 2*np.pi/(np.min([np.abs(de - loc_omega), loc_omega]))
```
It might be nice to comment on this formula, the use of np.min here seems non-obvious.

JanosJiri replied:
It is non-obvious. I use it because if I keep only `de - loc_omega`, then the numerical integral is extremely demanding when we are far from resonance. This is often the case when you have multiple states that are well separated. I will need to think about it a bit and come up with something better. Let's keep it like that and I will address at least part of it in a different pull request.

danielhollas commented on tests/test_pulse.py:
```diff
@@ -270,3 +270,50 @@ def test_field_envelope_shifted(make_pulse, envelope_type):
+            "gauss": 0.06628966759244047,
+            "lorentz": 0.44606065146635454,
+            "sech": 0.2328991679544244,
+            "sin": 1.4997597826618577e-35,
```
Somewhat surprising that the value for `sin` pulse is so much different.

Should we test for a larger set of `de` and `tprime` values?

JanosJiri reviewed (changes requested):
Nice work, thanks @danielhollas ! Everything looks good, I did one small commit fixing `field_envelopes.py` after wigner moved. Could you just expand the tests of the wigner for more times t' please? I added more extended comment as a reply to your comment.

JanosJiri commented:
One more important modification, I multiplied the integral in `wigner_transform` by a factor of 2. I didn't have it there before because it does not influence the result and might be redundant. However, it should be formally there and if we use analytic formulas for some envelopes later, there will be an inconsistency. Since I just created #27 , where I have this problem, I already changed it here because all the tests you will generate would be off by a factor of 2.

danielhollas commented on src/promdens/promdens.py:
```diff
@@ -163,6 +163,48 @@ def calc_field_envelope(self, t: np.ndarray) -> np.ndarray:
+        # instead of calculating the complex integral int_{-inf}^{inf}[E(t+s/2)E(t-s/2)exp(i(w-de)s)]ds we use the
+        # properties of even and odd fucntions and calculate 2*int_{0}^{inf}[E(t+s/2)E(t-s/2)cos((w-de)s)]ds
+        s = np.arange(0, factor[self.envelope_type]*self.fwhm, step=ds)
+        # Note: We assume here that de is in atomic units, otherwise it needs to be divided by hbar
```
I think it's good to mention this, I moved the comment up.

danielhollas commented:
> field_envelopes.py after Wigner moved. 

Thanks @JanosJiri for the fixes. I realized only in #26 that I need to update it. :-) In #26 I've added this script to CI so it should be less easy to break it.

I added more tprimes to the tests, please have a look if the numbers make sense to you. Feel free to push any modifications and merge, it's good from my side.

JanosJiri reviewed (approved):
Great, happy to merge!
